lib/opentelemetry/trace/batch_span_processor.lua

```diff
@@ -125,71 +76,75 @@ function _M.on_end(self, span)
         return
     end
 
+    -- Drop span if queue is full, otherwise add span to queue
     if self:get_queue_size() >= self.max_queue_size then
-        -- drop span
         if self.drop_on_queue_full then
             ngx.log(ngx.WARN, "queue is full, drop span: trace_id = ", span.ctx.trace_id, " span_id = ", span.ctx.span_id)
             self.dropping_count = self.dropping_count + 1
-            return
         end
-
-        -- export spans
-        process_batches_timer(self, self.batch_to_process)
-        self.batch_to_process = {}
-    end
-
-    table.insert(self.queue, span)
-    if #self.queue == 1 then
-        self.first_queue_t = now()
+    else
+        table.insert(self.queue, span)
+        if #self.queue == 1 then
+            self.first_queue_t = now()
+        end
     end
 
-    if #self.queue >= self.max_export_batch_size then
-        table.insert(self.batch_to_process, self.queue)
+    -- Make a batch if batch timeout has been reached or queue >= batch size
+    if now() - self.first_queue_t >= self.batch_timeout and #self.queue > 0 then
+        table.insert(self.batches_to_process, self.queue)
+        self.queue = {}
+    elseif #self.queue >= self.max_export_batch_size then
+        table.insert(self.batches_to_process, self.queue)
         self.queue = {}
     end
 
-    if not self.is_timer_running then
-        create_timer(self, self.inactive_timeout)
-    end
-end
+    -- Export if we've got enough batches. We want to export multiple batches
+    -- simultaneously to reduce the number of ngx.timer.at calls, since they
+    -- incur overhead.
+    if #self.batches_to_process >= self.maximum_pending_batches then
```

Comment:
If there was small spans, they will never be export.
I think we should keep the `is_timer_running ` design, there was no data race with `is_timer_running`, because lua threads in one nginx worker is running sequentially.

Reply:
If we keep `is_timer_running`, that means we can only ever have one export happening at a time, right?